src/jtag3.c

```diff
@@ -1685,44 +1707,48 @@ int jtag3_open_common(PROGRAMMER *pgm, const char *port) {
   }
 #endif
   if (rv < 0) {
-    // Check if SNAP or PICkit4 is in PIC mode
+    // Check if SNAP or PICkit4 are in PIC mode
     for(LNODEID ln=lfirst(pgm->id); ln; ln=lnext(ln)) {
-      if (str_starts(ldata(ln), "snap")) {
-        pinfo.usbinfo.vid = USB_VENDOR_MICROCHIP;
-        pinfo.usbinfo.pid = USB_DEVICE_SNAP_PIC_MODE;
-        int pic_mode = serial_open(port, pinfo, &pgm->fd);
-        if(pic_mode < 0) {
-          // Retry with alternative USB PID
-          pinfo.usbinfo.pid = USB_DEVICE_SNAP_PIC_MODE_ALT;
-          pic_mode = serial_open(port, pinfo, &pgm->fd);
-        }
-        if(pic_mode >= 0) {
-          msg_error("\n");
-          pmsg_error("MPLAB SNAP in PIC mode detected!\n");
-          imsg_error("Use MPLAB X or Microchip Studio to switch to AVR mode\n\n");
-          return -1;
-        }
-      } else if(str_starts(ldata(ln), "pickit4")) {
+      if (str_starts(ldata(ln), "snap") || str_starts(ldata(ln), "pickit4")) {
+        bool is_snap_pgm = str_starts(ldata(ln), "snap");
         pinfo.usbinfo.vid = USB_VENDOR_MICROCHIP;
-        pinfo.usbinfo.pid = USB_DEVICE_PICKIT4_PIC_MODE;
+        pinfo.usbinfo.pid = is_snap_pgm? USB_DEVICE_SNAP_PIC_MODE: USB_DEVICE_PICKIT4_PIC_MODE;
+        const int bl_pid = is_snap_pgm? USB_DEVICE_SNAP_PIC_MODE_BL: USB_DEVICE_PICKIT4_PIC_MODE_BL;
+        const char *pgmstr = is_snap_pgm? "MPLAB SNAP": "PICkit 4";
+        const unsigned char exit_bl_cmd[] = {0xe6};
+        const unsigned char enter_avr_mode_cmd[] = {0xf0, 0x01};
+        const unsigned char reset_cmd[] = {0xed};
+
         int pic_mode = serial_open(port, pinfo, &pgm->fd);
         if(pic_mode < 0) {
-          // Retry with alternative USB PID
-          pinfo.usbinfo.pid = USB_DEVICE_PICKIT4_PIC_MODE_ALT;
+          // Retry with bootloader USB PID
+          pinfo.usbinfo.pid = bl_pid;
           pic_mode = serial_open(port, pinfo, &pgm->fd);
         }
         if(pic_mode >= 0) {
           msg_error("\n");
-          pmsg_error("PICkit4 in PIC mode detected!\n");
-          imsg_error("Use MPLAB X or Microchip Studio to switch to AVR mode\n\n");
-          return -1;
+          pmsg_error("%s in %s mode detected\n",
+            pgmstr, pinfo.usbinfo.pid == bl_pid? "bootloader": "PIC");
+          if(mode_switch == PK4_SNAP_MODE_AVR) {
+            imsg_error("switching to AVR mode\n");
+            if(pinfo.usbinfo.pid == bl_pid)
+              serial_send(&pgm->fd, exit_bl_cmd, sizeof(exit_bl_cmd));
+            else {
+              serial_send(&pgm->fd, enter_avr_mode_cmd, sizeof(enter_avr_mode_cmd));
+              usleep(250*1000);
+              serial_send(&pgm->fd, reset_cmd, sizeof(reset_cmd));
+            }
+            imsg_error("please run Avrdude again to continue the session\n\n");
+          } else
+            imsg_error("use -xmode=avr to enter AVR mode\n\n");
+          return LIBAVRDUDE_SOFTFAIL;
```

Comment:
Why return with SOFTFAIL? Is the caller able to do sth with that?
 
The impression I got is that after `-xmode=avr` or `-xmode=pic` AVRDUDE should exit. Why not *document* that so (`-xmode=avr` or `-xmode=pic` switches the PICkit4 or Snap programmers into the  corresponding mode and exits) and `exit(0)`?

Come to think of this, why not create a function `switch_mode()` or some such that switches modes and call this function when the user issues `-xmode=...` This way the switching code does not have to reside in `open()` as there is no intention to open and continue. That also might(!) not require a new parameter in the `common_open()` function.

Reply:
> Why not document that 

I'm using `exit(0)` now. Do you mean documenting it in the docs?

> Come to think of this, why not create a function switch_mode()

Well, it _could_ be done so that `jtag3_open_common` returned known values if in PIC or bootloader mode, and did the switching outside. That's probably the cleanest solution. I'll look into it.

Reply:
I spent some time trying to mode the switching into a dedicated function so that `jtag3_open_common` can be left untouched. It turned out to be more complicated than I first thought, and I don't think it's worth my time. The current solution works really well and is easy to understand and follow (IMO).